# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('http/.../datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()
data.head()

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)
data.sample(7)

children  days_employed  dob_years education  education_id  \
16211         5     387.317579         35   среднее             1   
14374         2     939.283615         37   среднее             1   
6180          0    3645.887512         50    высшее             0   
9911          1     231.363347         32   среднее             1   
14185         1     558.705053         28    высшее             0   
2102          0    1449.253158         40   среднее             1   
14824         1     120.756402         29    высшее             0   

          family_status  family_status_id gender  income_type  debt  \
16211  гражданский брак                 1      F  госслужащий     0   
14374   женат / замужем                 0      F    компаньон     0   
6180    женат / замужем                 0      F    сотрудник     0   
9911    женат / замужем                 0      F    компаньон     0   
14185   женат / замужем                 0      F    компаньон     0   
2102          в разводе                 3      F    сотрудник     0   
14824   женат / замужем                 0      F    компаньон     0   

       total_income                     purpose total_income_category  
16211        126102       на проведение свадьбы                     C  
14374        187021                ремонт жилью                     C  
6180         167625  строительство недвижимости                     C  
9911         217083                 образование                     B  
14185        168302                  автомобиль                     C  
2102         103974                ремонт жилью                     C  
14824        143176                  автомобиль                     C

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)
data.sample(7)

children  days_employed  dob_years education  education_id  \
15199         0    2879.094521         25   среднее             1   
2203          0    1618.911339         56    высшее             0   
20874         0    3235.885835         58   среднее             1   
20556         0  364644.395227         62   среднее             1   
12056         1     753.375962         30   среднее             1   
19067         1  347168.436064         54   среднее             1   
18338         0    3776.681439         62   среднее             1   

          family_status  family_status_id gender  income_type  debt  \
15199   женат / замужем                 0      F  госслужащий     0   
2203   гражданский брак                 1      M    компаньон     0   
20874   женат / замужем                 0      F    сотрудник     0   
20556   женат / замужем                 0      F    пенсионер     0   
12056   женат / замужем                 0      M    сотрудник     0   
19067   женат / замужем                 0      F    пенсионер     0   
18338   женат / замужем                 0      M    сотрудник     0   

       total_income                           purpose total_income_category  \
15199        214393        покупка жилой недвижимости                     B   
2203         188402             на проведение свадьбы                     C   
20874         90752             получение образования                     C   
20556        262517     получение высшего образования                     B   
12056         66736  строительство жилой недвижимости                     C   
19067        157691                      ремонт жилью                     C   
18338        204528              покупка недвижимости                     B   

               purpose_category  
15199  операции с недвижимостью  
2203         проведение свадьбы  
20874     получение образования  
20556     получение образования  
12056  операции с недвижимостью  
19067  операции с недвижимостью  
18338  операции с недвижимостью

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
data_children = data.pivot_table(index = ['children'], values = 'debt', aggfunc = {'count', 'sum'})
data_children = data_children.rename(columns = {'count' : 'credits_count', 'sum' : 'overdue'})
data_children['share'] = round(data_children['overdue'] / data_children['credits_count'] * 100, 3)
data_children = data_children.sort_values(by = 'share', ascending = False)
display(data_children)

credits_count  overdue  share
children                               
4                    41        4  9.756
2                  2052      194  9.454
1                  4808      444  9.235
3                   330       27  8.182
0                 14091     1063  7.544
5                     9        0  0.000

**Вывод:** Наличие детей повышает вероятность задолжености. Вероятно, это связано с дополнительными расходами на детей. Однако, можно заметить, что в семьях где от 3 до 5 детей уровень должников ниже, чем в семьях с одним или двумя детьми. Это может быть связано либо с тем, что существуют программы поддержки многодетных семей или количество заемщиков в этих группах для данного сравнения недостаточно. Для категории граждан без детей самый низкий показатель должников.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
data_family_status = data.pivot_table(index = ['family_status'], values = 'debt', aggfunc = {'count', 'sum'})
data_family_status = data_family_status.rename(columns = {'count' : 'credits_count', 'sum' : 'overdue'})
data_family_status['share'] = round(data_family_status['overdue'] / data_family_status['credits_count'] * 100, 3)
data_family_status = data_family_status.sort_values(by = 'share', ascending = False)
display(data_family_status)

credits_count  overdue  share
family_status                                       
Не женат / не замужем           2796      273  9.764
гражданский брак                4134      385  9.313
женат / замужем                12261      927  7.561
в разводе                       1189       84  7.065
вдовец / вдова                   951       63  6.625

**Вывод:** семейное положение влияет на возврат кредита в срок. В группах "Не женат / не замужем" и "Гражданский брак" практически одинаково высокие показатели, а в остальных группах они значительно ниже.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
data_total_income = data.pivot_table(index = ['total_income_category'], values = 'debt', aggfunc = {'count', 'sum'})
data_total_income = data_total_income.rename(columns = {'count' : 'credits_count', 'sum' : 'overdue'})
data_total_income['share'] = round(data_total_income['overdue'] / data_total_income['credits_count'] * 100, 3)
data_total_income = data_total_income.sort_values(by = 'share', ascending = False)
display(data_total_income)

credits_count  overdue  share
total_income_category                               
E                                 22        2  9.091
C                              15921     1353  8.498
A                                 25        2  8.000
B                               5014      354  7.060
D                                349       21  6.017

**Вывод:** Уровень дохода влияет на возврат кредита. В группах "А" и "Е" очень мало данных их можно не учитывать, так как группе "А" кредитные средства с их уровнем дохода редко нужны, а группе "Е" с их низким уровнем доходов редко выдают кредиты. Сравнивая группы "В", "С" и "D" делаем вывод, что наибольший риск стать должниками у категории "C", а наименьший у группы "D".

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
data_purpose_category = data.pivot_table(index = ['purpose_category'], values = 'debt', aggfunc = {'count', 'sum'})
data_purpose_category = data_purpose_category.rename(columns = {'count' : 'credits_count', 'sum' : 'overdue'})
data_purpose_category['share'] = round(data_purpose_category['overdue'] / data_purpose_category['credits_count'] * 100, 3)
data_total_income = data_total_income.sort_values(by = 'share', ascending = False)
display(data_purpose_category)

credits_count  overdue  share
purpose_category                                       
операции с автомобилем             4279      400  9.348
операции с недвижимостью          10751      780  7.255
получение образования              3988      369  9.253
проведение свадьбы                 2313      183  7.912

**Вывод:** Граждане, берущие кредиты на получение образования и операции с автомобилем больше подвержены риску стать должниками. Граждане, которые берут кредит для проведения операций с недвижимостью более отвественно подходят к его погашению.

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Причинами могут быть: неверный ввод данных, сокрытие информации, повреждение файла данных при сохранении, технические проблемы.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Возможно потому что медиана устойчива к выбросам и аномальным значениям.

### Шаг 4: общий вывод.

  В ходе ознакомления с таблицей были обнаружены пропуски, аномальные значения, дубликаты в некоторых данных. 
  
  **Пропуски** в столбце "total_income — данные о доходах". На сумму дохода сильнее всего влияет тип занятости, поэтому заполнялись пропуски в этом столбце медианным значением по каждому типу из столбца "income_type".  

  **Аномальные значения** были найдены в столбце "children - количество детей" ([ 1,  0,  3,  2, **-1**,  4, **20**,  5]). Исправленно удалением строк.  
  
  **Дубликаты** (неявные) были выявлены в столбце "education - уровень образования" . В этом столбце были одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Все приведены к нижнему регистру.  
  
  **Категоризация данных**  
  Для улучшения и удобства просмотра была проведена категоризация и создан столбец "total_income_category" с категориями: 0–30000 — 'E'; 30001–50000 — 'D'; 50001–200000 — 'C'; 200001–1000000 — 'B'; 1000001 и выше — 'A'.
Распределяющий всех заёмщиков по уровню дохода в соответствующие категории.
  



1. Да, у заемщиков с детьми больше вероятность просрочки.  

  Группы где 1 или 2 ребёнка самые большие показатели (группа "1": количество заёмщиков 4808, количество просрочек 444, доля просроченных платежей 9.235), (группа "2": количество заёмщиков 2052, количество просрочек 194, доля просроченных платежей 9.454) Вэтом сегменте минимальные просрочки будут в группе "бездетные" или "0" детей (количество заёмщиков 14091, количество просрочек 1063, доля просрочек 7.544)
  
2. Да есть зависимость между семейным положением и возвратом кредита в срок.  

  В группах "Не женат/не замужем" и "гражданский брак" самые высокие показатели (Не женат / не замужем: количество заёмщиков 2796, количество просрочек 273, доля просроченных платежей 9.764), (гражданский брак: количество заёмщиков: 4134, количество просрочек 385, доля просроченных платежей	9.313). Самые минимальные показатели в группе "вдовец / вдова" (количество заёмщиков 951, количество просрочек 63, доля просроченных платежей 6.625).
  
3. Да есть зависимость между уровнем дохода и возвратом кредита в срок.  

  В группах "А", "D" и "Е" очень мало данных их можно не учитывать, так как группе "А" (количество заёмщиков 25, количество просрочек 2, доля просроченных платежей 8.000) кредитные средства с их уровнем дохода редко нужны, а группе "Е" (количество заёмщиков 22, количество просрочек 2, доля просроченных платежей 9.091) с их низким уровнем доходов редко выдают кредиты. В группе "D" (количество заёмщиков 349, количество просрочек 21, доля просроченных платежей 6.017) скорее всего таже ситуация что и в группе "Е". Сравнивая группы "В" И "С" делаем вывод, что наибольший риск стать должниками у категории "C" (количество заёмщиков 15921, количество просрочек 1353, доля просроченных платежей 8.498) , а наименьший у группы "В".  
  
4. Да разные цели кредита влияют на его возврат в срок.  

  Граждане, берущие кредиты на получение образования (количество заёмщиков 3988, количество просрочек 369, доля просроченных платежей 9.253) и операции с автомобилем (количество заёмщиков 4279, количество просрочек 400, доля просроченных платежей 9.348) больше подвержены риску стать должниками. Граждане, которые берут кредит для проведения операций с недвижимостью (количество заёмщиков 10751, количество просрочек 780, доля просроченных платежей 7.255) более отвественно подходят к его погашению.
 
Причинами появления пропусков в данных могут быть: неверный ввод данных, сокрытие информации, повреждение файла данных при сохранении, технические проблемы.

Заполнение пропусков для количественных переменных медианой является лучшим решением из-за устойчивости медианы к выбросам и аномальным значениям.